### DISCUSSION TITLE
(Possible, but unlikely) 2.0 Ideas

### DISCUSSION THREAD

**Discussion by GregoryConrad** [GregoryConrad/rearch-dart#234] [General]
As time has progressed, I've realized that a few API decisions I have made along the way were somewhat subpar. _If_ I were to do a 2.0 release, which I still am reluctant to do, I wanted to keep a tracking list of the possible changes for such a release since I have ideas for breaking changes time-to-time.

Most of these would be unlikely to break any codebases using ReArch unless they are using relatively niche features.

- `SideEffectApi.rebuild` should have a signature like:
  - `mutation` is now required
  - `cancelRebuild` no longer exists (just ensure the data returned by the side effect is `==` instead)
```dart
abstract interface class SideEffectRegistrar {
  // ...
  void rebuild(
    void Function() mutation,
  );
}
```
- Memoize the getter returned by `use.data` (and friends) when the data doesn't change
  - In lieu of `cancelRebuild` as described above
  - Yes, the capsule will rebuild, but this isn't a problem since the data returned will be the same and it'll use the `==` optimization
    - Hmm, but perhaps this would be an issue since folks just returning a tuple of callbacks would be creating new callbacks, and thus not `==`
    - Maybe we can suggest using `capsule.data` and similar and then have users make wrappers around that with new capsules?
- Remove `Option`
  - Instead, `AsyncValue` should just have nullable fields.
  - If users have nullable data, they should use their own `Option` type.
- https://github.com/GregoryConrad/rearch-dart/issues/302#issuecomment-3005206105 — use.data variants should return a new type (a final class), ValueWrapper, instead of a record

**Comment by satvikpendem** [comment]
You definitely shouldn't fear breaking changes because it's better to do them earlier on with fewer users than later on with likely many more users, just my two cents.
